burn-dataset/src/vision/image_folder.rs

```diff
@@ -0,0 +1,382 @@
+use crate::transform::{Mapper, MapperDataset};
+use crate::{Dataset, InMemDataset};
+
+use globwalk::{self, DirEntry};
+use image::{self, ColorType};
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::path::{Path, PathBuf};
+
+const SUPPORTED_FILES: [&str; 4] = ["bmp", "jpg", "jpeg", "png"];
+
+/// Image data type.
+#[derive(Debug, Clone, PartialEq)]
+pub enum PixelDepth {
+    /// 8-bit unsigned.
+    U8(u8),
+    /// 16-bit unsigned.
+    U16(u16),
+    /// 32-bit floating point.
+    F32(f32),
+}
+
+impl TryFrom<PixelDepth> for u8 {
+    type Error = &'static str;
+
+    fn try_from(value: PixelDepth) -> Result<Self, Self::Error> {
+        if let PixelDepth::U8(v) = value {
+            Ok(v)
+        } else {
+            Err("Value is not u8")
+        }
+    }
+}
+
+impl TryFrom<PixelDepth> for u16 {
+    type Error = &'static str;
+
+    fn try_from(value: PixelDepth) -> Result<Self, Self::Error> {
+        if let PixelDepth::U16(v) = value {
+            Ok(v)
+        } else {
+            Err("Value is not u16")
+        }
+    }
+}
+
+impl TryFrom<PixelDepth> for f32 {
+    type Error = &'static str;
+
+    fn try_from(value: PixelDepth) -> Result<Self, Self::Error> {
+        if let PixelDepth::F32(v) = value {
+            Ok(v)
+        } else {
+            Err("Value is not f32")
+        }
+    }
+}
+
+/// Image target for different tasks.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ImageTarget {
+    /// Image-level label.
+    Label(usize),
+    /// Object bounding box.
+    BoundingBox(BoundingBox),
+    /// Segmentation mask.
+    SegmentationMask(SegmentationMask),
+}
+
+/// Segmentation mask target.
+/// For semantic segmentation, a mask has a single channel (C = 1).
+/// For instance segmentation, there may be multiple masks per image (C >= 1).
+#[derive(Debug, Clone, PartialEq)]
+pub struct SegmentationMask {
+    /// Segmentation mask.
+    pub mask: Vec<usize>,
+}
+
+/// Object detection bounding box target.
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
+pub struct BoundingBox {
+    /// Coordinates.
+    pub coords: [f32; 4],
+
+    /// Box class label.
+    pub label: usize,
+}
+
+/// Image dataset item.
+#[derive(Debug, Clone, PartialEq)]
+pub struct ImageDatasetItem {
+    /// Image as a vector with a valid image type.
+    pub image: Vec<PixelDepth>,
+
+    /// Target for the image.
+    pub target: ImageTarget,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ImageDatasetItemRaw {
+    /// Image path.
+    pub image_path: PathBuf,
+
+    /// Image target.
+    /// The target string can be a category name or path to annotation file.
+    pub target: String,
```

Comment:
If the target can be parsed into all `ImageTarget` I would avoid using a string, but consider it as a byte array `Vec<u8>`, so that it is explicit that we need to decode it. If it's only used for classification, I would change the type to `ImageClassificationDatasetItemRaw`.

Reply:
Also, for classification, we may have an enum for the classification target, since a lot of time it is the label, but other time it can be the index. For the raw part we should support both ways, but when preprocessed, I think we should only expose the index.

Reply:
Agreed, this was supposed to target only classification. Will rename to `ImageClassificationDatasetItemRaw`. Also totally forgot about the index which could be provided instead of the string.

Reply:
Also, regarding `ImageTarget` I just realized it doesn't generalize very well for object detection. I think I will rename it to `Annotation` instead.